### ISSUE TITLE
Issues with libraries on 18.1.0

### ISSUE THREAD

**Issue by donqui** [zeromq/pyzmq#1331]
Hi,

I was using pyzmq 17.0.0 on my system that I previously installed with `pip install pyzmq==17.0.0`.

After I updated zeromq from I found that 18.0.0 installed with `pip install pyzmq==18.1.0` does not work on my system, but if I install it with `pip install pyzmq==18.1.0 --install-option="--zmq=bundled"` it starts working fine again.

While debugging I noticed that both 17.0.0 and 18.1.0 with --install-option use my systems libsodium, but that 18.1.0 installed without the option comes with it's own bundled libsodium.

*17.0.0*
```
└─ $ ▶ lsof -p xxxx | grep -iE 'libzmq|sodium'
python 31353 /usr/lib/x86_64-linux-gnu/libsodium.so.18.0.1
python 31353 /home/user/.virtualenvs/indy-node/lib/python3.5/site-packages/zmq/.libs/libzmq-0576c57a.so.5.0.2
```

*18.1.0*
```
└─ $ ▶ lsof -p xxxx | grep -iE 'libzmq|sodium'
python  /usr/lib/x86_64-linux-gnu/libsodium.so.18.0.1
python  /home/user/.virtualenvs/indy-node/lib/python3.5/site-packages/zmq/.libs/libsodium-bcf9f097.so.23.3.0
python 31445 /home/user/.virtualenvs/indy-node/lib/python3.5/site-packages/zmq/.libs/libzmq-1358af2c.so.5.2.2

└─ $ ▶ ldd libzmq-1358af2c.so.5.2.2
 linux-vdso.so.1 => (0x00007ffc07be3000)
 libsodium-bcf9f097.so.23.3.0 => /home/user/.virtualenvs/indy-node/lib/python3.5/site-packages/zmq/.libs/././libsodium-bcf9f097.so.23.3.0 (0x00007fc327e4a000)
 librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fc327c42000)
 libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc327a25000)
 libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc3276a3000)
 libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc32739a000)
 libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc326fd0000)
 libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc326dba000)
 /lib64/ld-linux-x86-64.so.2 (0x00007fc328447000)
```

*18.1.0 --install-option="–zmq=bundled"*
```
└─ $ ▶ lsof -p xxxx | grep -iE 'libzmq|sodium'
python /usr/lib/x86_64-linux-gnu/libsodium.so.18.0.1
python /home/user/.virtualenvs/indy-node/lib/python3.5/site-packages/zmq/libzmq.cpython-35m-x86_64-linux-gnu.so

└─ $ ▶ ldd libzmq.cpython-35m-x86_64-linux-gnu.so
 linux-vdso.so.1 => (0x00007ffd68cfd000)
 librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f03c8852000)
 libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f03c84d0000)
 libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f03c81c7000)
 libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f03c7fb1000)
 libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f03c7d94000)
 libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f03c79ca000)
 /lib64/ld-linux-x86-64.so.2 (0x00007f03c8d19000)
```

I do not know if this is expected, and if this is causing my issues, so I wanted to ask you guys if there are some gotchas that I need to know when updating.

**Comment by minrk** [comment]
When you say "it does not work" what does that mean? Do you have any LD_ environment variables like LD_LIBRARY_PATH set? It definitely should work to install the official wheel if you have libsodium on your system, so it's a problem on our side if that's not the case.

A Dockerfile or similar to reproduce the issue would be a big help.

> While debugging I noticed that both 17.0.0 and 18.1.0 with --install-option use my systems libsodium, but that 18.1.0 installed without the option comes with it's own bundled libsodium.

Yeah, that's expected. `--zmq=bundled` bundles *libzmq*, but not libsodium, but it can still link to a libsodium on the system, if present at build time. The official wheels are *not* built with `--zmq=bundled`, instead they do standard installs of libsodium and libzmq and use [auditwheel](https://github.com/pypa/auditwheel) to bundle the libraries.

**Comment by donqui** [comment]
Well, I have an app that uses pyzmq. I have libsodium installed on my system and I do have appropriate LD_ paths set.

My application works with 17.0.0 and with 18.1.0 when installed with `--zmq=bundled`, but when I used a 18.1.0 installed regularly without the install option my system starts lagging, like things do not reach the intended destination or get blocked somwhere.

I am honestly having trouble pinpointing the issue as I do not know why it would work with `--zmq=bundled` and not without it. Maybe you can shed some light on the differences because I the only differences I noticed are that libzmq is different between the two, and that one uses local libsodium while the other one does not.

**Comment by minrk** [comment]
It's a little suspect that it looks like Python is loading *both* the host libsodium and the bundled one. That doesn't seem right, and may point to the issue. Are you using something that's loading libsodium separately in the same process?

**Comment by donqui** [comment]
No. It works without a system lib installed.

I tried to fool it, so I linked the local version to the system one, it did not complain but it did not help.

**Comment by minrk** [comment]
To clarify: if you `apt-get remove` (or whatever) libsodium and change nothing else, everything works fine? But then reinstalling libsodium restores the performance problem?

Can I ask what your LD variables are? Setting LD variables can easily cause problems with library loading, and should ~never be required if software is properly installed and linked.

**Comment by donqui** [comment]
Uf that was a bad comment.

What I mean to say there is that pyzmq 18.1.0 without the --zmq=bundled works without a system lib available.

And generally I do not have any LD_* vars set, the only ones I had is when I tried to trick my app to always load only the local version of libsodium. Linking the local version to the system one did not help - there was only one version loaded but the app was still stuck.

And I lied to you, I now found that I am loading libsodium in different part of my app directly. Removing the previously set LD_* var caused a import failure.

Can I somehow force pyzmq to use the system lib instead of the local one?
